src/dune_engine/action_exec.ml

```diff
@@ -355,6 +357,19 @@ let rec exec t ~ectx ~eenv =
     Io.write_lines target (String.Set.to_list lines);
     Fiber.return Done
   | No_infer t -> exec t ~ectx ~eenv
+  | Using_terminal t ->
+    Fiber.Mutex.with_lock terminal_lock (fun () ->
+        Console.lock ();
+        let* result =
+          exec t ~ectx
+            ~eenv:
+              { eenv with
+                stdout_to = Process.Io.stdout
+              ; stderr_to = Process.Io.stderr
+              }
+        in
+        Console.unlock ();
```

Comment:
What happens when the `exec` above raises? Shouldn't we still unlock?

Reply:
Good catch.

Reply:
Good catch, fixed.